### ISSUE TITLE
Next step of Windows UWP support

### ISSUE THREAD

**Issue by MouriNaruto** [openssl/openssl#9125]
Continue the discussion on https://github.com/openssl/openssl/pull/8917.

We have some arguments to solve.

- What about taking UWP target support back to 1.1.1 as well? (I hope OpenSSL can take it back because there are a lot of projects that are waiting on this.)
- What about the Windows 10 SDK Build 17134 or earlier support? (Need to write "You also should add "no-asm" option when you use the Windows 10 SDK Build 17134 or earlier." on the note)
- What progress about the "Configure" script improvement? (Because we don't need to specify "no-uplink" at configure time.)
- What about provding a batch script for helping people compile?

Kenji Mouri

**Comment by MouriNaruto** [comment]
@levitte @mattcaswell @jdm

**Comment by mattcaswell** [comment]
> What about taking UWP target support back to 1.1.1 as well? (I hope OpenSSL can take it back because there are a lot of projects that are waiting on this.)

Adding new platforms is a new feature. New features are not added to stable branches because of the very fact that they are stable. Adding features is a destabilising action. Very often we find when new features are added we don't get it right straight away and there is some churn in the code as we fix stuff. IMO this cannot be backported to 1.1.1.,

> What about the Windows 10 SDK Build 17134 or earlier support? (Need to write "You also should add "no-asm" option when you use the Windows 10 SDK Build 17134 or earlier." on the note)

We should not be advocating no-asm for production use. Not only are there significant performance implications, there are security implications too. IMO we should not support earlier builds if we cannot get asm to work on them. My preferred route here would be to fix whatever the underlying cause of the asm failures is.

**Comment by levitte** [comment]
Regarding `no-asm` and `no-uplink`, I have an idea that would solve the problem with specifying them in a `disable` attribute.

See, the information they affect are currently in form of config attributes and special code in Configure. However, that information really belongs in build.info files, just like all other file and macro information.
Now that we have variables in build.info files, it's not even a hard task to move that information... Just tedious.

**Comment by MouriNaruto** [comment]
> Adding new platforms is a new feature. New features are not added to stable branches because of the very fact that they are stable.

I wonder what is the next LTS version after the OpenSSL 1.1.1?

**Comment by MouriNaruto** [comment]
> See, the information they affect are currently in form of config attributes and special code in Configure. However, that information really belongs in build.info files, just like all other file and macro information. Now that we have variables in build.info files, it's not even a hard task to move that information... Just tedious.

Well, I will try. But I am also curious about your experimental branch. (You have said "This can be remedied with late evaluation (I have an experimental PR that I didn't finish), maybe now's the time to introduce it...." before.)

**Comment by levitte** [comment]
For the idea I mentioned here, that experimental branch won't be needed.

**Comment by levitte** [comment]
#9166

**Comment by levitte** [comment]
> - What about the Windows 10 SDK Build 17134 or earlier support? (Need to write "You also should add "no-asm" option when you use the Windows 10 SDK Build 17134 or earlier." on the note)

If someone tells me how to detect the SDK version, I might be able to craft something.

**Comment by MouriNaruto** [comment]
@levitte 

In https://stackoverflow.com/questions/2665755, we can know the NTDDI_WIN10_RS5 macro has been defined in Windows 10 SDK Build 17763 (RS5) or later. So, write code like this:
```
#include <sdkddkver.h>
#if !defined(NTDDI_WIN10_RS5)
    #error Windows 10.0.17763.0 SDK is required
#endif
```

In https://github.com/Chuyu-Team/MINT/blob/565dc12ef9d3fd9c725187588c680b12cc403df4/MINT.h#L7539, you will notice something like the following thing.
> #if (NTDDI_VERSION >= NTDDI_WIN10_RS5)

I hope these can help you.

Kenji Mouri

**Comment by illera88** [comment]
@levitte any update on this? Any ETA? 

Thanks

**Comment by illera88** [comment]
@levitte are you alive? It has been 22 days now...

**Comment by levitte** [comment]
Hey, sorry for the delay, there's much going on.

@MouriNaruto, thanks for the details.  I worry that `sdkddkver.h` might not always be available, would you (or @illera88?) care to shed some light on that?

**Comment by illera88** [comment]
`sdkddkver.h` has been present since  Visual C++ 2010 Express so unless you support a 10 year old VC version I don't thing you will have any problem.

Do you need anything else?

**Comment by MouriNaruto** [comment]
I think `sdkddkver.h` has been present since at least Windows 8 SDK.
Windows Universal Platform has been present since Windows 10.
We should use Windows 10 SDK and Visual Studio 2015 or later if we need to develop an UWP app. 
So we don't need to worry about that.

**Comment by levitte** [comment]
Ah, so as long as I only check this for UWP builds, we're safe.  Thanks.

**Comment by levitte** [comment]
I ended up checking for the registry key that holds the SDK version.  Please try #9440 to see if it works for you.

**Comment by levitte** [comment]
#9440 has been merged (bcc0025).  I find it a little unfortunate that it went completely uncommented by those who wanted it, but feel free to try it out now.

**Comment by levitte** [comment]
So, I've seen no comment here since the fix.  I assume all is fine.  If not, raise a new issue.

**Comment by jdm** [comment]
I have been successfully using UWP-enabled openssl, so thank you!
